### DISCUSSION TITLE
How do dynamic `/api` routes (specifically catch all routes) work?

### DISCUSSION THREAD

**Discussion by qodesmith** [vercel/community#947] [Help] [answered]
I can't seem to get "catch all" routes to work (i.e. `[...slug]`). Here's the link to my [repo](https://github.com/qodesmith/vercel-api-test) which is a basic Vite-generated app with the added `/api` folder for serverless functions. As you'll see, I've got an `/api` folder structure as follows:

<img width="107" alt="Screen Shot 2022-10-11 at 7 02 58 AM" src="https://user-images.githubusercontent.com/11034792/195073750-574b22b8-8c8f-47f4-a3d4-81abfdd2a394.png">

For the "catch all" route of `[...slug].ts`, I'm simply sending back a response of `req.query`:
```typescript
// [...slug].ts
export default function handler(req: VercelRequest, res: VercelResponse) {
  res.json(req.query)
}
```

With that being said, I get the following responses from the api:
|url|response|route type|was expected|expected instead
|----|----|----|:----:|----|
|[/api/users](https://vercel-api-test-flax.vercel.app/api/users)|`{users: [...]}`|explicit|✅|-|
|[/api/users/1](https://vercel-api-test-flax.vercel.app/api/users/1)|`{user: {...}}`|dynamic|✅|-|
|[/api/cars](https://vercel-api-test-flax.vercel.app/api/cars)|404|-|✅|-|
|[/api/cars/corvette](https://vercel-api-test-flax.vercel.app/api/cars/corvette)|`{"...slug": "corvette"}`|catch all|❌|`{slug: ["corvette"]}`|
|[/api/cars/corvette/2017](https://vercel-api-test-flax.vercel.app/api/cars/corvette/2017)|404|catch all|❌|`{slug: ["corvette", "2017"]}`|

With a file name of `/api/cars/[...slug].ts`, I [expected](https://nextjs.org/docs/api-routes/dynamic-api-routes) Vercel to parse this as a "catch all" route and respond to any number of path segments from `/api/cars/unlimited/segments/here/...`. What am I doing wrong?

**Comment by qodesmith** [comment]
Could this be a bug in how Vercel parses the `/api` directory?

**Comment by amyegan** [comment, marked as answer]
Hi @qodesmith! The documentation you linked to is for Next.js, but your project does not seem to be a Next.js project. The [catch all API route](https://nextjs.org/docs/api-routes/dynamic-api-routes#catch-all-api-routes) behavior is a Next.js feature. 

I think it's possible build something to achieve a similar effect with Vite and other frameworks, but it would require some complex customization. I don't know enough about Vite to be able to offer specific advice on that.

**Reply by qodesmith** [reply]
Hi @amyegan, thanks for the reply. The Vercel docs mention in the [Deploying Serverless Functions](https://vercel.com/docs/concepts/functions/serverless-functions#deploying-serverless-functions) section that you need to add an `/api` directory in order to deploy functions. A little further down that page is the [Path Segments](https://vercel.com/docs/concepts/functions/serverless-functions#path-segments) section that seems to indicate it supports dynamic routes. Technically it only mentions dynamic routes with single brackets - `[name].ts`. This made me think it had parity with how it's done in Next.js since it's all deployed on Vercel infrastructure anyway.

Could you confirm if this is the only dynamic feature available on Vercel for the filesystem-based `/api` back end routes on Serverless Functions?

**Reply by amyegan** [reply]
Right, dynamic routes in non-Next apps are limited to just the single dynamic path segment value (or more if you string them together). The catch all feature belongs to Next.js. That's why you're getting `{"...slug": "corvette"}` instead of `{"slug": "corvette"}` back from [/api/cars/corvette](https://vercel-api-test-flax.vercel.app/api/cars/corvette).

If you know how many segments you expect, you can set up the path with multiple segments that will all pass through to the function at the end of that path. For example, I was able to get `{"second":"beta","first":"alpha"}` from path `/api/alpha/beta` by adding `/api/[first]/[second].ts` to my app.

<img width="135" alt="image" src="https://user-images.githubusercontent.com/8247105/195676404-0a1c9182-e09a-46f4-b617-d520d79af140.png">

**Reply by qodesmith** [reply]
Ok awesome. Thank you for confirmation! If you can, pass along this info as a feature request to Vercel to have `/api` feature parity with how Next.js does it.
